stl/inc/format

```diff
@@ -366,58 +366,347 @@ private:
     ptrdiff_t _Next_arg_id = 0;
 };
 
-// TODO: test coverage
+enum class _Basic_format_arg_type : uint8_t {
+    _None,
+    _Int_type,
+    _UInt_type,
+    _LLong_type,
+    _ULLong_type,
+    _Bool_type,
+    _Char_type,
+    _Float_type,
+    _Double_type,
+    _LDouble_type,
+    _Pointer_type,
+    _CString_type,
+    _String_type,
+    _Custom_type,
+};
+static_assert(static_cast<int>(_Basic_format_arg_type::_Custom_type) <= 16);
+
 template <class _Context>
 class basic_format_arg {
 public:
-    class handle;
+    using _CharType = typename _Context::char_type;
 
-private:
-    using _Char_type = typename _Context::char_type;
-
-    using _Format_arg_value = variant<monostate, int, unsigned, long long, unsigned long long, bool,
-        typename _Context::char_type, float, double, long double, const void*, const typename _Context::char_type,
-        basic_string_view<typename _Context::char_type>, handle>;
-
-    _Format_arg_value _Value;
-
-public:
     class handle {
     private:
         const void* _Ptr;
-        void (*_Format)(basic_format_parse_context<_Char_type>& _Parse_ctx, _Context _Format_ctx, const void*);
+        void (*_Format)(basic_format_parse_context<_CharType>& _Parse_ctx, _Context _Format_ctx, const void*);
         friend basic_format_arg;
 
     public:
-        void format(basic_format_parse_context<_Char_type>& _Parse_ctx, _Context& _Format_ctx) {
+        template <class _Ty>
+        explicit handle(const _Ty& _Val) noexcept
+            : _Ptr(_STD addressof(_Val)),
+              _Format([](basic_format_parse_context<_CharType>& _Parse_ctx, _Context& _Format_ctx, const void* _Ptr) {
+                  typename _Context::template formatter_type<_Ty> _Formatter;
+                  _Parse_ctx.advance_to(_Formatter.parse(_Parse_ctx));
+                  _Format_ctx.advance_to(_Formatter.format(*static_cast<const _Ty*>(_Ptr), _Format_ctx));
+              }){};
+
+        void format(basic_format_parse_context<_CharType>& _Parse_ctx, _Context& _Format_ctx) {
             _Format(_Parse_ctx, _Format_ctx, _Ptr);
         }
     };
 
-    basic_format_arg() noexcept = default;
+    // TRANSITION, LLVM-49072
+    basic_format_arg() noexcept : _Active_state(_Basic_format_arg_type::_None), _No_state() {}
+
+    explicit basic_format_arg(const int _Val) noexcept
+        : _Active_state(_Basic_format_arg_type::_Int_type), _Int_state(_Val) {}
+    explicit basic_format_arg(const unsigned int _Val) noexcept
+        : _Active_state(_Basic_format_arg_type::_UInt_type), _UInt_state(_Val) {}
+    explicit basic_format_arg(const long long _Val) noexcept
+        : _Active_state(_Basic_format_arg_type::_LLong_type), _LLong_state(_Val) {}
+    explicit basic_format_arg(const unsigned int long long _Val) noexcept

```

Comment:
There are several occurrences of `unsigned int long long` that should be `unsigned long long`.